design/baremetal-operator/image-builder-integration.md

```diff
@@ -0,0 +1,241 @@
+request. The baremetal-operator should watch the image request and recreate it
+when this occurs. Since `ValidateManagementAccess()` runs in every state, the
+ramdisk image URL will be updated as soon as a new one is available, and the
+next boot of the ramdisk will make use of the new value.
+
+### Risks and Mitigations
+
+If the consumer of the BareMetalHost resource is able to modify the
+networkData, they will be able to influence the configuration of the Host after
+they have deprovisioned and potentially released it. This can be mitigated by
+not rebuilding the image once it is created. To allow for rebuilding when
+desired an administrator can delete the request, but it is comparatively easy
+to withold RBAC access to do so from other classes of users.
+
+### Work Items
+
+- Implement the new `RamDiskImageRequest` CRD
+- Implement a basic controller to reconcile the `RamDiskImageRequest`
+- Add runtime flag to configure image building integration
+- Add a watch on the `RamDiskImageRequest` for the `BareMetalHost` reconciler
+- Create the `RamDiskImageRequest` when it is missing
+- Pause registration until the image URL is available
+- Pass the ramdisk URL to Ironic
+
+### Dependencies
+
+None
+
+### Test Plan
+
+Enable the functionality in the integration test, using the default (trivial)
+controller implementation for the `RamDiskImageRequest`. This should produce
+the same result as today, but exercise the integration between the the two
+resource types.
+
+### Upgrade / Downgrade Strategy
+
+This functionality will be disabled by default, so that existing users don't
+need to do anything special on upgrade.
+
+### Version Skew Strategy
+
+The API surface area is very small, so we are unlikely to encounter version
+skew issues.
+
+## Drawbacks
+
+This adds complexity inside the baremetal-operator, for something that is not
+needed by many users.
+
+## Alternatives
+
+### Have Ironic customise the image
+
+Currently Ironic just boots the Node with a given kernel and initrd. If Ironic
+were able to supply the networkData to the Node as well when booting the
+ramdisk image, then customisation could be performed within the ramdisk. It
+appears that theoretically Ironic can already [pass the networkData in this
+way](https://docs.openstack.org/ironic/latest/admin/dhcp-less.html#configuring-network-data)
+when creating an ISO to boot the deploy image using virtualmedia. This should
+be enough to accomplish our goals here. It is implemented in the iLO and
+redfish drivers, and inherited by idrac-redfish-virtualmedia.
+
+One downside of this is that there is no access from outside for a user to
+request their own image build (e.g. to manually boot a server into the deploy
+ramdisk for discovery).
+
+With this strategy we would always have to update the network data when it
+changes; we have no way of distinguishing between changes made by an
+administrator or other user.
```

Comment:
We should point out that the customization approaches for different agent image types (especially coreos) aren't necessarily compatible with ironic.